Add memory fences to kernel language (#1379)
alpaka-group/alpaka: docs/source/basic/library.rst

They are parametrized by the origin of the calculation as well as the unit in which the values are calculated.
For example, ``alpaka::getWorkDiv<alpaka::Grid, alpaka::Threads>(acc)`` returns a vector with the extents of the grid in units of threads.

Memory fences
`````````````

> **Note**: Memory fences should not be mistaken for synchronization functions between threads. They solely enforce the
> ordering of certain memory operations (see below) and how other threads observe this order. If you need to rely on the
> results of memory operations being visible to other threads please use ``alpaka::syncBlockThreads`` or atomic functions
> instead.

Reviewer: I don't think the wording is fully clear here, because a fence *is* used to influence the order of how changes from one thread become visible to another thread. I would add some emphasis:

```suggestion
**Note**: Memory fences should not be mistaken for synchronization functions between threads.
They solely enforce the ordering of certain memory operations (see below) and restrict the possibilities of how other threads *can* observe this order.
If you need to guarantee that the results of memory operations are visible to other threads you must use ``alpaka::syncBlockThreads`` or atomic functions instead.
```

Author: I don't really see a difference between those sentences, sorry. I'm also not certain that fences restrict "the possibilities of how other threads can observe this order" since there aren't any possibilities left once a fence is issued.

Reviewer: >I'm also not certain that fences restrict "the possibilities of how other threads can observe this order" since there aren't any possibilities left once a fence is issued.

Of course there are possibilities:

```
int a = 0, b = 0;
void thread1() {
    a = 1;
    fence();
    b = 3;
}
void thread2() {
    read(a);
    fence();
    read(b);
}
```

Thread 2 can observe a=0/b=0 or a=1/b=0 or a=1/b=3. However, the possibility of a=0/b=3 is not possible because of the fence. The fence thus restricts the 4 possible variable assignments to 3. The behavior of thread 2 is not deterministic by the fence. That is also why I wrote that thread 2 **can** observe the behavior, because in the case of a=0/b=0 it does not observe the changes from thread 1.

Author: You're talking about other threads observing different **results**, but the **order of instructions** is still the same. With the fence it is guaranteed that a is written before b, but the observed results can be different.

Reviewer: I see. Then please rephrase at least:
```suggestion
**Note**: Memory fences should not be mistaken for synchronization functions between threads. They solely enforce the
ordering of certain memory instructions (see below) and restrict how other threads can observe this order. If you need to rely on the
results of memory operations being visible to other threads please use ``alpaka::syncBlockThreads`` or atomic functions
instead.
```
This way we are specific about instructions. And a fence does not enforce how other threads observe results, it merely restricts what observations are possible.

Some accelerators (like GPUs) follow weaker cache coherency rules than x86 CPUs. In order to avoid storing to (or loading
from) a cache or register it is necessary to prefix all observed buffers with `ALPAKA_DEVICE_VOLATILE`. This enforces
that all loads / stores access the actual global / shared memory location.

Example:

.. code-block:: cpp

    /* Initial values: 
     * vars[0] = 1
     * vars[1] = 2
     */
    template<typename TAcc>
    ALPAKA_FN_ACC auto operator()(TAcc const& acc, bool* success, ALPAKA_DEVICE_VOLATILE int* vars) const -> void
    {
        auto const idx = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc)[0u];

        // Global thread 0 is producer
        if(idx == 0)
        {
            vars[0] = 10;
            alpaka::mem_fence(acc, alpaka::memory_scope::Device{});
            vars[1] = 20;
        }

        auto const b = vars[1];
        alpaka::mem_fence(acc, alpaka::memory_scope::Device{});
        auto const a = vars[0];

        /* Possible results at this point:
         * a == 1 && b == 2
         * a == 10 && b == 2
         * a == 10 && b == 20
         *
         * but NOT:
         * a == 1 && b == 20
         */

>     }

Reviewer: I am starting to think that this code is undefined in C++, as it is missing an atomic operation. See: https://en.cppreference.com/w/cpp/atomic/atomic_thread_fence

> Fence-fence synchronization
> A release fence FA in thread A synchronizes-with an acquire fence FB in thread B, if
> There exists an atomic object M,

This code does not have an atomic, and volatile does replace this behavior.

Author: Ugh, I thought I could do this without having to call another atomic operation (on godbolt the correct CPU instructions are generated by `std::atomic_thread_fence`), but you're right.
